Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp

```diff
@@ -0,0 +1,268 @@
+	"cutoff when conducting per formula selection.\n"
+	"e.g. 0.1 = organisms in the top 90% are kept. use 0.0 for classic Lexicase.");
+std::shared_ptr<ParameterLink<bool>> LexicaseOptimizer::epsilonByRangePL =
+Parameters::register_parameter("OPTIMIZER_LEXICASE-epsilonByRange", false,
+	"if true epsilon will be relative to min and max score"
+	"\n  i.e. keep orgs with (score > maxScore - (maxScore-minScore) * epsilon)"
+	"\nif false, epsilon will be relative to organism ranks"
+	"\n  i.e. keep (best current keepers * epsilon) organisms");
+
+std::shared_ptr<ParameterLink<int>> LexicaseOptimizer::tournamentSizePL = 
+Parameters::register_parameter("OPTIMIZER_LEXICASE-poolSize", -1,
+	"number of organisms used when selecting parent(s) in the lexicase algorithm, -1 indicates to use entire population");
+
+std::shared_ptr<ParameterLink<bool>> LexicaseOptimizer::recordOptimizeValuesPL = 
+Parameters::register_parameter("OPTIMIZER_LEXICASE-recordOptimizeValues", true,
+	"record optimize values to data files using optimizeFormulaNames");
+
+
+LexicaseOptimizer::LexicaseOptimizer(std::shared_ptr<ParametersTable> PT_)
+    : AbstractOptimizer(PT_) {
+
+	std::vector<std::string> optimizeFormulasStrings;
+	convertCSVListToVector(optimizeFormulasPL->get(PT), optimizeFormulasStrings);
+
+	for (auto s : optimizeFormulasStrings) {
+		optimizeFormulasMTs.push_back(stringToMTree(s));
+	}
+
+	// get names to use with scores
+	if (optimizeFormulaNamesPL->get(PT) == "default") {
+		// user has not defined names, auto generate names
+		for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
+			scoreNames.push_back("optimizeValue_" + std::to_string(fIndex));
+		}
+	}
+	else {
+		// user has defined names, use those
+		convertCSVListToVector(optimizeFormulaNamesPL->get(PT), scoreNames);
+	}
+
+
+	epsilon = epsilonPL->get(PT);
+	epsilonByRange = epsilonByRangePL->get(PT);
+	tournamentSize = tournamentSizePL->get(PT);
+	nextPopSizeFormula = stringToMTree(nextPopSizePL->get(PT));
+	numberParents = numberParentsPL->get(PT);
+	recordOptimizeValues = recordOptimizeValuesPL->get(PT);
+	
+	// leave this undefined so that max.csv is not generated
+	//optimizeFormula = optimizeValueMT;
+
+	popFileColumns.clear();
+	if (recordOptimizeValues){
+		for (auto name : scoreNames) {
+			popFileColumns.push_back(name);
+		}
+	}
+}
+
+
+int LexicaseOptimizer::lexiSelect(const std::vector<int> &tournamentIndexList) {
+	if (!scoresHaveDelta) { // if all scores are the same! pick random
+		return Random::getIndex(tournamentIndexList.size());
+	}
+
+	// generate a vector with formulasSize and fill with values from 0 to formulasSize in a random order.
+	std::vector<int> formulasOrder;
+	std::vector<int> sourceValues; // temp vector used to generate formulasOrder
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		sourceValues.push_back(i);
+	}
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		int pick = Random::getIndex(sourceValues.size());
```

Comment:
If Random:: provides its number generator, this could be a shuffle.

Reply:
done